File: 1807-evaluate-the-bracket-pairs-of-a-string/1807-evaluate-the-bracket-pairs-of-a-string.js
/**
 * @param {string} s
 * @param {string[][]} knowledge
 * @return {string}
 */
var evaluate = function(s, knowledge) {
    const map = new Map();
        for (const list of knowledge)
            map.set(list[0], list[1]);

        let sb = "";
        let l = 0;
        while (l < s.length) {
            if (s[l] != '(') 
                sb += s[l++];
            else {
                let r = l + 1;
                while (s[r] != ')') r++;

                const sub = s.substring(l + 1, r);
                sb += map.get(sub) ?? "?";
                
                l = r + 1;
            }
        }
        return sb;
};

File: 1807-evaluate-the-bracket-pairs-of-a-string/1807-evaluate-the-bracket-pairs-of-a-string.java
class Solution {
    public String evaluate(String s, List<List<String>> knowledge) {
        HashMap<String, String> map = new HashMap<>();
        for (List<String> list : knowledge)
            map.put(list.get(0), list.get(1));

        StringBuilder sb = new StringBuilder();
        int l = 0;
        while (l < s.length()) {
            if (s.charAt(l) != '(') 
                sb.append(s.charAt(l++));
            else {
                int r = l + 1;
                while (s.charAt(r) != ')') r++;

                String sub = s.substring(l + 1, r);

                map.put(sub, map.getOrDefault(sub, "?"));
                sb.append(map.get(sub));
                
                l = r + 1;
            }
        }
        return sb.toString();
    }
}